### PR TITLE
tnat64: init at 0.05

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change

It's an old software but it checks out.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested it works via `env LD_PRELOAD=$(nix-build -A tnat64 --no-out-link)/lib/tnat64/libtnat64.so nc <ip> <port>`
- [x] Tested execution of all binary files
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Note: careful with LD_PRELOAD: if your main glibc doesn't match the one in nixpkgs master, it will crash.

### CODE REVIEW

File: pkgs/tools/networking/tnat64/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{ lib, stdenv, fetchFromGitHub, autoreconfHook }:
```
Comment (drewrisinger) [inline pkgs/tools/networking/tnat64/default.nix:1]:
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, autoreconfHook
}:
```
Standard is one line per

File: pkgs/tools/networking/tnat64/default.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    homepage = "https://github.com/andrewshadura/tnat64";
+    license = licenses.gpl2Plus;
+    longDescription = ''
+      TNAT64 is an interceptor which redirects outgoing TCPv4 connections
+      through NAT64, thus enabling an application running on an IPv6-only host
+      to communicate with the IPv4 world, even if that application does not
+      support IPv6 at all.
+    '';
+    platforms = platforms.unix;
+    maintainers = [ maintainers.rnhmjoj ];
```
Comment (drewrisinger) [inline pkgs/tools/networking/tnat64/default.nix:33]: optional: 
```suggestion
    maintainers = with maintainers; [ rnhmjoj ];
```

**Review by drewrisinger — CHANGES_REQUESTED** [review]
* [x] Minor diff points
* [x] Commit LGTM
* [x] Builds via ``nix-review`` on x86-64-linux. Untested

```
1 package built:
tnat64
```

File: pkgs/tools/networking/tnat64/default.nix (hunk 1 above)
Comment (legendofmiracles) [reply pkgs/tools/networking/tnat64/default.nix:1]: It's fine if there are only so few

**Review by drewrisinger — APPROVED** [review]

**Comment by OPNA2608** [comment]
Fails to build on Darwin.

> ```
> make  install-exec-hook
> make[2]: Entering directory '/private/tmp/nix-build-tnat64-0.05.drv-0/source'
> cd /nix/store/yhxvz2pc9md03cbp0d4gsiw1w6psmrcp-tnat64-0.05/lib/tnat64 && \
>  rm -f libtnat64.a && \
>  rm -f libtnat64.la && \
>  rm -f libtnat64.so && \
>  rm -f libtnat64.so.0 && \
>  mv libtnat64.so.0.0.0 libtnat64.so
> mv: cannot stat 'libtnat64.so.0.0.0': No such file or directory
> make[2]: *** [Makefile:1218: install-exec-hook] Error 1
> ```

Please patch [that hook](https://github.com/andrewshadura/tnat64/blob/cfb19ac2e77287768e922244e8c2cb57a812f17e/Makefile.am#L41-L47) to work on Darwin (`libtnat64.0.dylib` is the longest variant I can find in the logs), patch it away altogether or mark the package broken on Darwin.

**Comment by rnhmjoj** [comment]
Uhm, now that i think about it I'm not sure it will work on darwin.
This program is hooking libc functions via the `ld.so` preloading mechanism (`LD_PRELOAD`), but darwin doesn't use `ld.so`. It seems there is a `DYLD_INSERT_LIBRARIES` that should do the same job, though.

@OPNA2608 are you avaiable for more testing? It'll try to patch it a bit before marking this broken.

**Comment by OPNA2608** [comment]
The log was taken from [OfBorg](https://github.com/NixOS/nixpkgs/pull/121714/checks?check_run_id=2502367393) but I do have a Darwin machine at home if you want me to test anything.

**Comment by rnhmjoj** [comment]
> I do have a Darwin machine at home if you want me to test anything.

Yes, that would be great. Thank you.

I've just pushed a commit to try fix the build and the `tnat64` shell script on darwin.

**Comment by OPNA2608** [comment]
I'm not sure if I'm doing it correctly but I don't think it works. When I run it on my NixOS system with your example command to a remote machine's SSH port it prints

```
20:32:46 libtnat64(49329): Could not open NAT64 configuration file (/etc/tnat64.conf), assuming all networks local
```

and exits.

Doing the same on my macOS system just runs the netcat command like normal without any such message, checking `lsof -f` shows `nc` only having IPv4 connections.

**Comment by rnhmjoj** [comment]
Ok, so it's building and running, good. It needs a minimal configuration file to work, try putting this into /etc/tnat64.conf
```
nat64_prefix = 64:ff9b::
```
If you don't want to touch /etc, you can use the `TNAT64_CONF_FILE` variable to change the path.
To check if the hooks are working try running the program with `strace -e socket nc 127.0.0.1 1234`. If it worked it should print something like
```
socket(AF_INET6, SOCK_STREAM|SOCK_NONBLOCK, IPPROTO_TCP) = 3
```

Note: it's important the program you're running (netcat) is built using the same version of libc as of tnat64. To test I did `nix-build -A netcat` and used `result-nc/bin/nc`.

**Comment by OPNA2608** [comment]
I can confirm that I get `AF_INET6` on my NixOS system. On my Darwin one running `nc` from Nixpkgs' `netcat-gnu` package (regular and `-openbsd` just exist immediately with return value `1`) doesn't seem to do anything different than running it without tnat64. `lsof -i` still says IPv4 and I don't know how to use `dtruss` to print any helpful syscall data like with `strace`.

**Comment by rnhmjoj** [comment]
I guess the only way is use a real NAT64 server. Try this config
```
nat64_prefix = 2a00:1098:2b::
```
and running `printf 'GET /plain HTTP/1.1\nHost: ipecho.net\n\n' | result/bin/tnat64 result-nc/bin/nc 34.117.59.81 80`.
If it worked it should print the IPv4 address of the NAT64 server (46.235.231.114, it seems), not yours.

**Pull request merged**